File: queu/leetcode-queu.js
// Runtime: 52 ms, faster than 73.39% of JavaScript online submissions for Implement Queue using Stacks.
// Memory Usage: 32.9 MB, less than 100.00% of JavaScript online submissions for Implement Queue using Stacks.
// https://leetcode.com/problems/implement-queue-using-stacks/submissions/
// author: @theasteve email: stevenaguilarcs@gmail.com


// Initialize your data structure here.
var MyQueue = function() {
  this.first = null;
  this.last = null;
  this.size = 0;
};

var Node = function(value) {
  this.val = value;
  this.next = null;
};

/**
 * Push element x to the back of queue. 
 * @param {number} x
 * @return {void}
 */
MyQueue.prototype.push = function(x) {
  // Add to the end of the list
  let oldLast = this.last;
  this.last = new Node(x);
  if(this.first === null) {
    this.first = this.last;
  } else {
    oldLast.next = this.last
  }
  this.size++;
  return
};

/**
 * Removes the element from in front of queue and returns that element.
 * @return {number}
 */
MyQueue.prototype.pop = function() {
  let item = this.first.val;
  this.first = this.first.next;
  if(this.first === null) {
    this.last = null;
  }
  this.size--;
  return item;
};

/**
 * Get the front element.
 * @return {number}
 */
MyQueue.prototype.peek = function() {
  return this.first.val;
};

/**
 * Returns whether the queue is empty.
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {
  return this.first === null;
};

/**
 * Returns whether the queue is empty.
 * @return {boolean}
 */
MyQueue.prototype.empty = function() {
  return this.first === null;
};

File: queu/leetcode_queu.rb
# Runtime: 32 ms, faster than 75.00% of Ruby online submissions for Implement Queue using Stacks.
# Memory Usage: 9.4 MB, less than 100.00% of Ruby online submissions for Implement Queue using Stacks.
# https://leetcode.com/problems/implement-queue-using-stacks/submissions/

class MyQueue

=begin
    Initialize your data structure here.
=end
    def initialize
      @first = nil
      @last = nil
      @size = 0
    end


=begin
    Push element x to the back of queue.
    :type x: Integer
    :rtype: Void
=end
    def push(x)
      old_last = @last
      @last = Node.new(x)
      if(@first == nil)
        @first = @last
      else
        old_last.next = @last
      end
      @size += 1
    end


=begin
    Removes the element from in front of queue and returns that element.
    :rtype: Integer
=end
    def pop
      item = @first.val
      @first = @first.next
      @last = nil if(@first == nil)
      @size -= 1
      item
    end


=begin
    Get the front element.
    :rtype: Integer
=end
    def peek
      @first.val
    end


=begin
    Returns whether the queue is empty.
    :rtype: Boolean
=end
    def empty
      @first.nil?
    end

    private

    class Node
      attr_accessor :val, :next

      def initialize(value)
        @val = value
        @next = nil
      end
    end
end
